File: pipecat/lib/sanitize.py
"""PII sanitization utilities.

Port of lib/sanitize.js — masks phone numbers, names, and truncates content
for safe logging.
"""

from __future__ import annotations

import re
import unicodedata


_PHONE_RE = re.compile(r"\b(?:\+?1[-.\s]?)?(?:\(?\d{3}\)?[-.\s]?)\d{3}[-.\s]?\d{4}\b")
_EMAIL_RE = re.compile(r"\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b", re.I)
_SSN_RE = re.compile(r"\b\d{3}[-.\s]?\d{2}[-.\s]?\d{4}\b")
_CONTROL_RE = re.compile(r"[\x00-\x08\x0B\x0C\x0E-\x1F\x7F-\x9F\u200B-\u200D\u202A-\u202E\u2066-\u2069\uFEFF]")
_SPACE_RE = re.compile(r"\s+")
_PROMPT_INJECTION_LINE_RE = re.compile(
    r"("
    r"\b(ignore|disregard|forget|override|bypass|negate|replace)\b.{0,100}"
    r"\b(previous|prior|above|earlier|system|developer|assistant|user|all)?\s*instructions?\b"
    r"|"
    r"\b(reveal|print|show|output|dump|leak|exfiltrate|send|email|text|transmit)\b.{0,100}"
    r"\b(system\s+prompt|developer\s+message|hidden\s+prompt|instructions?|api\s*key|secret|token|password|canary(?:[_-]?[a-z0-9]+)*)\b"
    r"|"
    r"\b(system|developer|assistant|user|tool)\s*(prompt|message|instructions?)\s*[:=-]"
    r"|"
    r"^\s*(?:#{1,6}\s*)?(system|developer|assistant|user|tool)\s*[:>-]"
    r"|"
    r"</?(system|developer|assistant|user|tool)\b[^>]*>"
    r"|"
    r"\bdo\s+not\s+(summarize|sanitize|filter|redact|remove)\b"
    r"|"
    r"\btrusted\s+(system|developer)\s+(message|instruction|prompt)\b"
    r")",
    re.I,
)


def mask_phone(phone: str | None) -> str:
    """Mask a phone number: '+15551234567' → '***4567'."""
    if not phone:
        return "[no-phone]"
    digits = re.sub(r"\D", "", phone)
    if len(digits) < 4:
        return "****"
    return "***" + digits[-4:]


def mask_contact_info(text: str | None) -> str:
    """Mask phone, email, and SSN-like substrings in free-form text."""
    if text is None:
        return ""
    value = str(text)
    value = _PHONE_RE.sub(lambda match: mask_phone(match.group(0)), value)
    value = _EMAIL_RE.sub("[email redacted]", value)
    return _SSN_RE.sub("[ssn redacted]", value)


def mask_name(name: str | None) -> str:
    """Mask a name for logs: 'David Zuluaga' → 'David Z.'."""
    if not name:
        return "[unknown]"
    parts = name.split()
    if len(parts) == 1:
        return parts[0]
    return parts[0] + " " + " ".join(p[0] + "." for p in parts[1:])


def truncate(text: str | None, max_len: int = 30) -> str:
    """Truncate content for safe logging."""
    if not text:
        return ""
    if len(text) <= max_len:
        return text
    return text[:max_len] + "..."


def sanitize_untrusted_text(
    text: str | None,
    *,
    max_len: int | None = 1000,
    replacement: str = "[removed unsafe instruction]",
    redact_contact_info: bool = True,
) -> str:
    """Remove prompt-control instructions from untrusted free-form text."""
    if text is None:
        return ""

    value = unicodedata.normalize("NFKC", str(text))
    value = _CONTROL_RE.sub("", value)
    if redact_contact_info:
        value = mask_contact_info(value)

    safe_lines: list[str] = []
    removed = False
    for raw_line in value.splitlines():
        line = raw_line.strip()
        if not line:
            continue
        if _PROMPT_INJECTION_LINE_RE.search(line):
            removed = True
            continue
        safe_lines.append(line)

    value = _SPACE_RE.sub(" ", " ".join(safe_lines)).strip()
    if not value and removed:
        value = replacement
    if max_len and len(value) > max_len:
        value = truncate(value, max_len)
    return value

File: lib/sanitize.js
/**
 * PII sanitization utilities.
 * Masks phone numbers, limits content previews, redacts sensitive fields.
 */

const PHONE_PATTERN = /(?:\+?1[\s.-]?)?(?:\(?\d{3}\)?[\s.-]?)\d{3}[\s.-]?\d{4}/g;
const EMAIL_PATTERN = /\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/gi;
const SSN_PATTERN = /\b\d{3}[-.\s]?\d{2}[-.\s]?\d{4}\b/g;
const CONTROL_PATTERN = /[\u0000-\u0008\u000B\u000C\u000E-\u001F\u007F-\u009F\u200B-\u200D\u202A-\u202E\u2066-\u2069\uFEFF]/g;
const SPACE_PATTERN = /\s+/g;

const PROMPT_INJECTION_LINE_PATTERNS = [
  /\b(ignore|disregard|forget|override|bypass|negate|replace)\b.{0,100}\b(previous|prior|above|earlier|system|developer|assistant|user|all)?\s*instructions?\b/i,
  /\b(reveal|print|show|output|dump|leak|exfiltrate|send|email|text|transmit)\b.{0,100}\b(system\s+prompt|developer\s+message|hidden\s+prompt|instructions?|api\s*key|secret|token|password|canary(?:[_-]?[a-z0-9]+)*)\b/i,
  /\b(system|developer|assistant|user|tool)\s*(prompt|message|instructions?)\s*[:=-]/i,
  /^\s*(?:#{1,6}\s*)?(system|developer|assistant|user|tool)\s*[:>-]/i,
  /<\/?(system|developer|assistant|user|tool)\b[^>]*>/i,
  /\bdo\s+not\s+(summarize|sanitize|filter|redact|remove)\b/i,
  /\btrusted\s+(system|developer)\s+(message|instruction|prompt)\b/i,
];

/**
 * Mask a phone number: "5551234567" -> "***4567"
 */
export function maskPhone(phone) {
  if (!phone) return '[no-phone]';
  const digits = phone.replace(/\D/g, '');
  if (digits.length < 4) return '****';
  return '***' + digits.slice(-4);
}

/**
 * Truncate content for safe logging: "long string..." -> "long str..."
 */
export function truncate(str, maxLen = 30) {
  if (!str) return '';
  const value = String(str);
  if (value.length <= maxLen) return value;
  return value.substring(0, maxLen) + '...';
}

/**
 * Mask a senior name for logs: "David Zuluaga" -> "David Z."
 */
export function maskName(name) {
  if (!name) return '[unknown]';
  const parts = name.split(' ');
  if (parts.length === 1) return parts[0];
  return parts[0] + ' ' + parts.slice(1).map(p => p[0] + '.').join(' ');
}

/**
 * Mask phone-looking substrings inside free-form error messages.
 */
export function maskPhonesInText(text) {
  if (text == null) return text;
  return String(text).replace(PHONE_PATTERN, match => maskPhone(match));
}

export function maskEmailsInText(text) {
  if (text == null) return text;
  return String(text).replace(EMAIL_PATTERN, '[email redacted]');
}

export function maskSensitiveText(text) {
  if (text == null) return text;
  return maskEmailsInText(maskPhonesInText(String(text))).replace(SSN_PATTERN, '[ssn redacted]');
}

function isUnsafePromptLine(line) {
  return PROMPT_INJECTION_LINE_PATTERNS.some(pattern => pattern.test(line));
}

/**
 * Remove prompt-control instructions from untrusted text before it can be sent
 * to users, vendors, or future LLM prompts.
 */
export function sanitizeUntrustedMessageText(text, options = {}) {
  if (text == null) return '';
  const {
    maxLen = 1000,
    replacement = '[removed unsafe instruction]',
    redactContactInfo = true,
  } = options;

  let value = String(text).normalize('NFKC').replace(CONTROL_PATTERN, '');
  if (redactContactInfo) {
    value = maskSensitiveText(value);
  }

  const safeLines = [];
  let removed = false;

  for (const rawLine of value.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line) continue;
    if (isUnsafePromptLine(line)) {
      removed = true;
      continue;
    }
    safeLines.push(line);
  }

  value = safeLines.join(' ').replace(SPACE_PATTERN, ' ').trim();
  if (!value && removed) {
    value = replacement;
  }
  if (maxLen && value.length > maxLen) {
    value = truncate(value, maxLen);
  }
  return value;
}

/**
 * Return a safe, non-PHI error summary for structured logs.
 */
export function sanitizeError(error) {
  if (!error) return null;
  return {
    name: error.name || 'Error',
    code: error.code,
    message: truncate(maskSensitiveText(error.message || String(error)), 160),
  };
}
